text/0011-landmarks.md

```diff
@@ -0,0 +1,99 @@
+This can be an empty list.
+
+```
+struct LandmarkObservation {
+  std::string id;
+  transform::Rigid3d transform;
+  double translation_weight;
+  double rotation_weight;
+};
+
+struct LandmarkData {
+  common::Time time;
+  std::vector<LandmarkObservation> landmarks;
+};
+```
+
+### Pose graph
+Pose graph will collect and store information about all observations provided by sensors.
+```
+mapping::MapById<mapping::LandmarkId, mapping::LandmarkNode>
+```
+where `LandmarkNode` contains all the global pose of the landmark and all landmark observations:
+
+```
+LandmarkNode {
+  LandmarkObservation {
+    common::Time time;
+    int trajectory_id;
+    
+    transform::Rigid3d landmark_to_tracking_transform;
+    double translation_weight;
+    double rotation_weight;
+  };
+  std::vector<LandmarkObservation> landmark_observations;
+  transform::Rigid3d global_landmark_pose;
+};
+```
+Note, that the `global_landmark_pose` is a vector of optimization parameters.
+Before the first sparse pose optimization run we can store extrapolated pose here instead and use it for rendering.
+Another option would be not populating the `global_landmark_pose` at all and not rendering the respective landmark on the map before optimization.
+
+
+### Optimization
+Landmarks will enter sparse pose graph optimization as additional independent variables.
+It increases the dimension of the optimization by `number_of_landmarks * number_of_parameters_per_landmark`, which could be undesirable when there is a significant number of landmarks.
+
+#### Cost function
+Landmark sensor data contains only the landmark position relative to the robot. In order to formulate a landmark-to-trajectory cost functional the global pose of the robot at the moment of the observation has to be known.
+Since a landmark observation can be made at the time between two trajectory nodes and, moreover, the global poses of the trajectory nodes are state variables as well, the translation and rotation of the landmark at the moment can be interpolated using the previous and the next node, i.e. taking a convex combination of the translation components and slerp of the quaternions.
+This logic is already implemented within `TransformInterpolationBuffer`.
+
+Residuals depend on the parameter block `[c_previous, c_next, c_landmark]`, i.e. on the global poses of the previous and next trajectory nodes and the global pose of the landmark itself:
+
+![component diagram](0011-assets/cost.png "Cost function")
+
+where `w` is the weight vector; `c` denotes global poses and `h` relative ones.
+
+There will be no loss function as for intra submap constraints.
+
+#### Constraints
+The cost function for landmark-to-trajectory constraints depends on the previous and next trajectory nodes.
+They are not known at the moment when the sensor data comes, but we can find them prior to the `RunOptimization` call using `AddWorkItem()` for landmark observations that were not yet processed.
```

Comment:
@cschuet Yes, this cannot be done outside `AddWorkItem()` for thread safety and lock contention. We do not have to do it prior to `RunOptimization()` though. This is the part I do not understand: Why not compute the previous and next trajectory node in the optimization problem and add the appropriate cost function? What would be the benefit of precomputing `LandmarkConstraint`s that warrants this additional complexity?